turbo/src/main/kotlin/dev/hotwire/turbo/errors/HttpError.kt

```diff
@@ -0,0 +1,154 @@
+            override val reasonPhrase = "Unprocessable Entity"
+        }
+
+        data object PreconditionRequired : ClientError {
+            override val statusCode = 428
+            override val reasonPhrase = "Precondition Required"
+        }
+
+        data object TooManyRequests : ClientError {
+            override val statusCode = 429
+            override val reasonPhrase = "Too Many Requests"
+        }
+
+        data class Other(
+            override val statusCode: Int,
+            override val reasonPhrase: String?
+        ) : ClientError
+    }
+
+    /**
+     * Errors representing HTTP server errors in the 500..599 range.
+     */
+    sealed interface ServerError : HttpError {
+        data object InternalServerError : ServerError {
+            override val statusCode = 500
+            override val reasonPhrase = "Internal Server Error"
+        }
+
+        data object NotImplemented : ServerError {
+            override val statusCode = 501
+            override val reasonPhrase = "Not Implemented"
+        }
+
+        data object BadGateway : ServerError {
+            override val statusCode = 502
+            override val reasonPhrase = "Bad Gateway"
+        }
+
+        data object ServiceUnavailable : ServerError {
+            override val statusCode = 503
+            override val reasonPhrase = "Service Unavailable"
+        }
+
+        data object GatewayTimeout : ServerError {
+            override val statusCode = 504
+            override val reasonPhrase = "Gateway Timeout"
+        }
+
+        data object HttpVersionNotSupported : ServerError {
+            override val statusCode = 505
+            override val reasonPhrase = "Http Version Not Supported"
+        }
+
+        data class Other(
+            override val statusCode: Int,
+            override val reasonPhrase: String?
+        ) : ServerError
+    }
+
+    companion object {
+        fun from(errorResponse: WebResourceResponse): HttpError {
+            return getError(errorResponse.statusCode, errorResponse.reasonPhrase)
+        }
+
+        fun from(statusCode: Int): HttpError {
+            return getError(statusCode, null)
+        }
+
+        private fun getError(statusCode: Int, reasonPhrase: String?): HttpError {
+            if (statusCode in 400..499) {
+                return ClientError::class.sealedSubclasses
+                    .mapNotNull { it.objectInstance }
+                    .firstOrNull { it.statusCode == statusCode }
+                    ?: ClientError.Other(statusCode, reasonPhrase)
+            }
+
+            if (statusCode in 500..599) {
+                return ServerError::class.sealedSubclasses
+                    .map { it.objectInstance }
+                    .firstOrNull { it?.statusCode == statusCode }
+                    ?: ServerError.Other(statusCode, reasonPhrase)
+            }
+
+            throw IllegalArgumentException("Invalid HTTP error status code: $statusCode")
```

Comment:
I'm wondering if we should be more resilient than this, even if the server gives us garbage...

Reply:
Yeah I wondered this too. I could maybe go with an `ClientError`, `ServerError` and  `UnknownError` at the top level. That'd be safe and not cause apps to crash if a server misbehaved.

Reply:
Addressed here https://github.com/hotwired/turbo-android/pull/317/commits/cc4cde3dd2ebe58b4c044bbfea6a901143006485